programs/protocol-contracts-solana/src/lib.rs

```diff
@@ -393,6 +428,49 @@ fn recover_eth_address(
     Ok(eth_address)
 }
 
+fn validate_signature_or_authority(
+    pda: &mut Account<Pda>,
+    authority: &Signer,
+    signature: [u8; 64],
+    recovery_id: u8,
+    message_hash: [u8; 32],
+    nonce: u64,
+    purpose: &str,
+) -> Result<()> {
+    // signature provided, recover and verify that tss is the signer
+    if signature != [0u8; 64] {
+        if nonce != pda.nonce {
+            msg!("mismatch nonce");
+            return err!(Errors::NonceMismatch);
+        }
+
+        let mut concatenated_buffer = Vec::new();
+        concatenated_buffer.extend_from_slice(purpose.as_bytes());
+        concatenated_buffer.extend_from_slice(&pda.chain_id.to_be_bytes());
+        concatenated_buffer.extend_from_slice(&nonce.to_be_bytes());
```

Comment:
shall we add the whitelist mint account to the hash here for good measure?

Reply:
also consider a domain separator like so 
https://github.com/zeta-chain/protocol-contracts-solana/blob/d0ecf1d2fe203660666f521465ba258cdce6372d/programs/protocol-contracts-solana/src/lib.rs#L302